patches/tModLoader/Terraria/ModLoader/GlobalType.cs

```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+
+namespace Terraria.ModLoader
+{
+	public abstract class GlobalType : ModType
+	{
+		internal ushort index;
+
+		/// <summary>
+		/// Whether to create a new instance of this Global for every entity that exists. 
+		/// Useful for storing information on an entity. Defaults to false. 
+		/// Return true if you need to store information (have non-static fields).
+		/// </summary>
+		public virtual bool InstancePerEntity => false;
+
+		internal GlobalType() { }
+
+		internal static T Instance<T>(Instanced<T>[] globals, ushort index) where T : GlobalType {
+			for (int i = 0; i < globals.Length; i++) {
+				var g = globals[i];
+
+				if (g.index == index) {
+					return g.instance;
+				}
+			}
+
+			return null;
+		}
+
+		internal static TResult GetGlobal<TEntity, TGlobal, TResult>(Instanced<TGlobal>[] globals, TResult baseInstance) where TGlobal : GlobalType<TEntity> where TResult : TGlobal
+			=> Instance(globals, baseInstance.index) as TResult ?? throw new KeyNotFoundException($"Instance of '{typeof(TResult).Name}' does not exist on the current {typeof(TEntity).Name}.");
+
+		internal static TResult GetGlobal<TEntity, TGlobal, TResult>(Instanced<TGlobal>[] globals, bool exactType) where TGlobal : GlobalType<TEntity> where TResult : TGlobal {
+			if (exactType) {
+				return GetGlobal<TEntity, TGlobal, TResult>(globals, ModContent.GetInstance<TResult>());
+			}
+
+			for (int i = 0; i < globals.Length; i++) {
+				if (globals[i].instance is TResult result) {
+					return result;
+				}
+			}
+
+			throw new KeyNotFoundException($"Instance of '{typeof(TResult).Name}' does not exist on the current {typeof(TEntity).Name}.");
+		}
+
+		internal static bool TryGetGlobal<TEntity, TGlobal, TResult>(Instanced<TGlobal>[] globals, TResult baseInstance, out TResult result) where TGlobal : GlobalType<TEntity> where TResult : TGlobal {
```

Comment:
This doesn't actually need the `TEntity` param here, as you can do everything with the non-generic `GlobalType` constraint. And similar elsewhere